File: src/prismeai/resources/conversations.py
from __future__ import annotations

from typing import Any, Optional, TYPE_CHECKING

from .._pagination import SyncCursorPage, AsyncCursorPage

if TYPE_CHECKING:
    from .._client import SyncAPIClient, AsyncAPIClient


class _BaseMixin:
    _workspace_id: str

    def _path(self, *segments: str) -> str:
        parts = "/".join(segments)
        return f"/workspaces/slug:{self._workspace_id}/webhooks/v1/{parts}"


class Conversations(_BaseMixin):
    def __init__(self, client: SyncAPIClient, workspace_id: str) -> None:
        self._client = client
        self._workspace_id = workspace_id

    def list(self, agent_id: str, *, page: Optional[int] = None, limit: Optional[int] = None) -> SyncCursorPage[Any]:
        return SyncCursorPage(self._client, "GET", self._path("agents", agent_id, "conversations"), page_size=limit or 20, start_page=page or 1)

    def create(self, agent_id: str, **kwargs: Any) -> Any:
        result = self._client.post(self._path("agents", agent_id, "conversations"), json=kwargs)
        # API returns contextId — normalize to id
        if isinstance(result, dict) and not result.get("id") and result.get("contextId"):
            result["id"] = result["contextId"]
        return result

    def get(self, agent_id: str, conversation_id: str) -> Any:
        return self._client.get(self._path("agents", agent_id, "conversations", conversation_id))

    def update(self, agent_id: str, conversation_id: str, **kwargs: Any) -> Any:
        return self._client.patch(self._path("agents", agent_id, "conversations", conversation_id), json=kwargs)

    def delete(self, agent_id: str, conversation_id: str) -> None:
        self._client.delete(self._path("agents", agent_id, "conversations", conversation_id))

    def share(self, agent_id: str, conversation_id: str, **kwargs: Any) -> None:
        self._client.post(self._path("agents", agent_id, "conversations", conversation_id, "share"), json=kwargs)


class AsyncConversations(_BaseMixin):
    def __init__(self, client: AsyncAPIClient, workspace_id: str) -> None:
        self._client = client
        self._workspace_id = workspace_id

    def list(self, agent_id: str, *, page: Optional[int] = None, limit: Optional[int] = None) -> AsyncCursorPage[Any]:
        return AsyncCursorPage(self._client, "GET", self._path("agents", agent_id, "conversations"), page_size=limit or 20, start_page=page or 1)

    async def create(self, agent_id: str, **kwargs: Any) -> Any:
        result = await self._client.post(self._path("agents", agent_id, "conversations"), json=kwargs)
        # API returns contextId — normalize to id
        if isinstance(result, dict) and not result.get("id") and result.get("contextId"):
            result["id"] = result["contextId"]
        return result

    async def get(self, agent_id: str, conversation_id: str) -> Any:
        return await self._client.get(self._path("agents", agent_id, "conversations", conversation_id))

    async def update(self, agent_id: str, conversation_id: str, **kwargs: Any) -> Any:
        return await self._client.patch(self._path("agents", agent_id, "conversations", conversation_id), json=kwargs)

    async def delete(self, agent_id: str, conversation_id: str) -> None:
        await self._client.delete(self._path("agents", agent_id, "conversations", conversation_id))

    async def share(self, agent_id: str, conversation_id: str, **kwargs: Any) -> None:
        await self._client.post(self._path("agents", agent_id, "conversations", conversation_id, "share"), json=kwargs)

File: src/resources/agents/conversations.ts
import { BaseResource } from '../../core/base-resource.js';
import type { HttpClient } from '../../core/http-client.js';
import type { PageIterator } from '../../core/pagination.js';
import type {
  Conversation,
  ConversationCreateParams,
  ConversationUpdateParams,
  ConversationListParams,
  ConversationShareParams,
} from '../../types/conversations.js';

export class Conversations extends BaseResource {
  private readonly workspaceSlug: string;

  constructor(httpClient: HttpClient, workspaceSlug: string) {
    super(httpClient);
    this.workspaceSlug = workspaceSlug;
  }

  private path(agentId: string, ...segments: string[]): string {
    return this.buildPath(this.workspaceSlug, 'agents', agentId, 'conversations', ...segments);
  }

  /** List conversations for an agent. */
  list(agentId: string, params?: ConversationListParams): PageIterator<Conversation> {
    return this._paginate<Conversation>(this.path(agentId), params);
  }

  /** Create a new conversation for an agent. */
  async create(agentId: string, params?: ConversationCreateParams): Promise<Conversation> {
    const result = await this._post<Conversation>(this.path(agentId), params);
    // API returns contextId — normalize to id
    if (!result.id && result.contextId) {
      result.id = result.contextId;
    }
    return result;
  }

  /** Get a conversation by ID. */
  get(agentId: string, conversationId: string): Promise<Conversation> {
    return this.httpClient.get<Conversation>(this.path(agentId, conversationId));
  }

  /** Update a conversation. */
  update(agentId: string, conversationId: string, params: ConversationUpdateParams): Promise<Conversation> {
    return this._patch<Conversation>(this.path(agentId, conversationId), params);
  }

  /** Delete a conversation. */
  delete(agentId: string, conversationId: string): Promise<void> {
    return this._del<void>(this.path(agentId, conversationId));
  }

  /** Share a conversation. */
  share(agentId: string, conversationId: string, params: ConversationShareParams): Promise<void> {
    return this._post<void>(this.path(agentId, conversationId, 'share'), params);
  }
}
